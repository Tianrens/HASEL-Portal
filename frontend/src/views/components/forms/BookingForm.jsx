import { React, useEffect, useState } from 'react';
import {
    Checkbox,
    Divider,
    FormControl,
    FormControlLabel,
    FormGroup,
    FormHelperText,
    FormLabel,
} from '@material-ui/core';
import dayjs from 'dayjs';
import styles from './BookingForm.module.scss';
import TextField from '../TextField/CustomTextField';
import GpuBookingGanttZoomable from '../gpuBookingGantt/GpuBookingGanttZoomable';

const getTimestamp = (date, time) => dayjs(`${date}T${time}:00`).valueOf();

const BookingForm = ({ updateBookingState, numGPUs, data }) => {
    const GPUList = Array.from(Array(numGPUs).keys());

    const [startDate, setStartDate] = useState('');
    const [endDate, setEndDate] = useState('');
    const [startTime, setStartTime] = useState('');
    const [endTime, setEndTime] = useState('');
    const [selectedGPUs, setGPUs] = useState(data?.gpuIndices ?? []);
    const noGPUSelected = selectedGPUs.length === 0;
    const timingInvalid = getTimestamp(startDate, startTime) > getTimestamp(endDate, endTime);

    const handleSelect = (event, GPU) => {
        if (event.target.checked) {
            setGPUs([...selectedGPUs, GPU]);
        } else {
            setGPUs(selectedGPUs.filter((gpu) => gpu !== GPU));
        }
    };
    useEffect(() => {
        if (data) {
            setStartDate(dayjs(data.startTimestamp).format('YYYY-MM-DD'));
            setStartTime(dayjs(data.startTimestamp).format('HH:mm'));
            setEndDate(dayjs(data.endTimestamp).format('YYYY-MM-DD'));
            setEndTime(dayjs(data.endTimestamp).format('HH:mm'));
            setGPUs(data.gpuIndices);
        }
    }, [data]);

    useEffect(() => {
        updateBookingState(
            {
                startTimestamp: getTimestamp(startDate, startTime),
                endTimestamp: getTimestamp(endDate, endTime),
                gpuIndices: selectedGPUs,
            },
            noGPUSelected || timingInvalid,
        );
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [startDate, startTime, endDate, endTime, selectedGPUs]);

    return (
        <>
            <form
                id='booking-form'
                autoComplete='off'
                className={styles.form}
                onSubmit={(event) => {
                    event.preventDefault();
                }}
            >
                <div className={styles.inputContainer}>
                    <TextField
                        title='Start Date'
                        type='date'
                        value={startDate}
                        setValue={setStartDate}
                        error={timingInvalid}
                        helperText={timingInvalid ? 'Start time must be before end time' : ''}
                    />
                    <TextField
                        title='Start Time'
                        type='time'
                        value={startTime}
                        error={timingInvalid}
                        setValue={setStartTime}
                    />
                </div>
                <div className={styles.inputContainer}>
                    <TextField
                        title='End Date'
                        error={timingInvalid}
                        type='date'
                        value={endDate}
                        setValue={setEndDate}
                        // For nicer alignment on desktop mode
                        helperText={timingInvalid ? ' ' : ''}
                    />
                    <TextField
                        title='End Time'
                        error={timingInvalid}
                        type='time'
                        value={endTime}
                        setValue={setEndTime}
                    />
                </div>
                <div className={styles.inputContainer}>
                    <FormControl required error={noGPUSelected}>
                        <FormLabel>Select GPUs</FormLabel>
                        <FormGroup row>
                            {GPUList.map((GPU) => (
                                <FormControlLabel
                                    control={
                                        <Checkbox
                                            onClick={(event) => handleSelect(event, GPU)}
                                            checked={selectedGPUs.includes(GPU)}
                                        />
                                    }
                                    label={`GPU ${GPU}`}
                                    key={GPU}
                                />
                            ))}
                        </FormGroup>
                        {noGPUSelected && <FormHelperText>Select at least 1 GPU</FormHelperText>}
                    </FormControl>
                </div>
            </form>
            <Divider className={styles.divider} />
            <div className={styles.gantt}>
                <GpuBookingGanttZoomable />
            </div>
            <Divider className={styles.divider} />
        </>
    );
};

export default BookingForm;
