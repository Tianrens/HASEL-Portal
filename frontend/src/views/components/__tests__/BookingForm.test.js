import React from 'react';
import renderer from 'react-test-renderer';
import StylesProvider from '@mui/styles/StylesProvider';
import BookingForm from '../forms/BookingForm';

Object.defineProperty(global.self, 'crypto', {
    value: {
        // eslint-disable-next-line no-unused-vars
        getRandomValues: (arr) => null,
    },
});

test('BookingForm renders properly', () => {
    const snapshotComponent = renderer.create(
        <StylesProvider injectFirst>
            <BookingForm numGPUs={5} />
        </StylesProvider>,
    );
    const tree = snapshotComponent.toJSON();
    expect(tree).toMatchSnapshot();
});
