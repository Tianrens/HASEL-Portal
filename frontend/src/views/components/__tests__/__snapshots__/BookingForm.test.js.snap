// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm renders properly 1`] = `
Array [
  <form
    autoComplete="off"
    className="form"
    id="booking-form"
    onSubmit={[Function]}
  >
    <div
      className="inputContainer"
    >
      <div
        className="container"
      >
        <p
          className="title"
        >
          Start Date
          <span
            className="asterisk"
          >
             *
          </span>
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={true}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="date"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
          <p
            className="MuiFormHelperText-root MuiFormHelperText-contained Mui-error Mui-required"
          >
            Start time must be after current time
          </p>
        </div>
      </div>
      <div
        className="container"
      >
        <p
          className="title"
        >
          Start Time
          <span
            className="asterisk"
          >
             *
          </span>
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={true}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="time"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      className="inputContainer"
    >
      <div
        className="container"
      >
        <p
          className="title"
        >
          End Date
          <span
            className="asterisk"
          >
             *
          </span>
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={true}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="date"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
          <p
            className="MuiFormHelperText-root MuiFormHelperText-contained Mui-error Mui-required"
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </p>
        </div>
      </div>
      <div
        className="container"
      >
        <p
          className="title"
        >
          End Time
          <span
            className="asterisk"
          >
             *
          </span>
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={true}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="time"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      className="inputContainer"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormLabel-root Mui-error Mui-required"
        >
          Select GPUs
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk Mui-error"
          >
            â€‰
            *
          </span>
        </label>
        <div
          className="MuiFormGroup-root MuiFormGroup-row"
        >
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-8"
                  data-indeterminate={false}
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              GPU 0
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-8"
                  data-indeterminate={false}
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              GPU 1
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-8"
                  data-indeterminate={false}
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              GPU 2
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-8"
                  data-indeterminate={false}
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              GPU 3
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-8"
                  data-indeterminate={false}
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              GPU 4
            </span>
          </label>
        </div>
        <p
          className="MuiFormHelperText-root Mui-error Mui-required"
        >
          Select at least 1 GPU
        </p>
      </div>
    </div>
  </form>,
  <hr
    className="MuiDivider-root divider"
  />,
  <div
    className="gantt"
  />,
  <hr
    className="MuiDivider-root divider"
  />,
]
`;
